### ISSUE TITLE
Is it possible to use this core as an USB-UART bridge?

### ISSUE THREAD

**Issue by francis2tm** [no2fpga/no2muacm#3]
Hello,
Amazing work with this usb core. Is it possible to use it directly as a USB-UART bridge? Like D+ and D- entering the FPGA and (UART) TX & RX leaving via output pins

**Comment by smunaut** [comment]
You need to wire it up to an actual uart core like those in no2misc to convert from the internal 'axi stream like' interface and actual uart tx/rx. It's pretty trivial to do for a fixed speed. If you want to support changing speed (and/or other control signals) you'd have to add support for the appropriate control requests for that. Not very complicated and should be fairly small logic wise too.

If you want help hooking that up, join the 1bitsquared discord server and ping me there in the #fpga channel.

**Comment by francis2tm** [comment]
Thanks a lot for the support, I'm going to try for myself first ;)

**Comment by francis2tm** [comment]
Hello again!
I'm starting now to work on this USB-UART bridge based on this usb core. Do you think it's possible to generate a DTR and a RTS signals? Essencially I want to use an ECP5 to act like a USB-UART bridge to interface a ESP32 with a computer. It'd be amazing if it'd be possible to flash firmware via the USB-UART bridge to the ESP32 too! And the ESP32 can use DTR and RTS signals to put itself it bootloader mode, like so:
![image](https://user-images.githubusercontent.com/43780565/183259909-1074df7c-8557-4d9d-bd1b-78d99a6e9167.png)
Do you think it's possible?
Thanks a lot ;)

**Comment by smunaut** [comment]
Yes, it's possible, you should be able to easily add a couple of GPIO and process the control request to drive them.

Note that ATM the core has some ice40 specific things in it that will need adapting for ecp5. Nothing major (just IOBs and BRAM) and I actually have in a branch an ecp5 version that you could probably patch in easily, but something to consider.

Any reason to use an ECP5 rather than an ice40 ?  If that's the only thing you want to put in there, it seems overkill.

**Comment by francis2tm** [comment]
Thanks a lot, I'm using an ECP5 cuz I have a full SoC inside it, unfortunately the ice40 doesn't have resources for everything :|
